### DISCUSSION TITLE
Parse Postgres array values in Core

### DISCUSSION THREAD

**Discussion by alecov** [sqlalchemy/sqlalchemy#9774] [Usage Questions] [answered]
I'm using sqlalchemy's core expressions to work with a PostgreSQL database, eg.:

```python
import sqlalchemy as sql

for row in map(lambda x: x._asdict(),
    connection.execute(sql.select(sql.text("*")
    .select_from(sql.table("some_table"))))
):
    # do something with row as dict:
    print(row["id"])
```

This works great since the returned rows are correctly typed, but when a PostgreSQL array is involved, the returned element in the row comes up as a wrapped string:

```python
for row in map(lambda x: x._asdict(),
    connection.execute(sql.select(sql.text("*")
    .select_from(sql.table("some_table"))))
):
    # <class 'str'>
    print(type(row["array_col"]))
    # {"(1,\"Item 1\")","(2,\"Item 2\")","(3,\"Item 3\")"}
    print(row["array_col"])
```

How do I parse this string? I get this is possible with ORM and `ARRAY` but I wonder whether I can do this with core alone.

BTW: is `sql.text("*")` the best way to select for `*`?

Thanks.

**Comment by CaselIT** [comment]
Hi,

This highly dependents on the driver/dbapi in use. Some will automatically parse array, while some do not.

You can specify the type of the columns in the text cluase, by using columns: https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.TextClause.columns

If you know what columns are in the tables I would create `table()` objects with the columns and types.
Or you can use the inspector to load the table columns with their types dynamically https://docs.sqlalchemy.org/en/20/core/reflection.html#sqlalchemy.engine.reflection.Inspector

**Reply by alecov** [reply]
Hi Case, thanks for the reply.

The actual problem seems to be that in my real table the array element is of a composite type (and it's not related to arrays at all, so please scrap that):

```sql
create type my_point as (x int, y int);
create table my_table (id int primary key, point my_point);
insert into my_table values (1, row(1,2)::my_point), (2, row(3,4)::my_point);
```

```python
print(sql.inspect(engine).get_columns("my_table"))
# SAWarning: Did not recognize type 'my_point' of column 'point'
```

Glancing at https://docs.sqlalchemy.org/en/20/orm/composites.html# it seems SA does not support composite types outside declarative/ORM.

What would you suggest to do in this case?

**Reply by CaselIT** [reply, marked as answer]
composite is not the custom PG type. It's just a python construct, whose saved in different column.

Custom PG types are not supported at the moment. The issue tracking this is https://github.com/sqlalchemy/sqlalchemy/issues/2450

I'm not sure if there is a solution here, other than maybe checking if the driver you are using has direct support for it. @zzzeek do you have suggestions?

**Reply by alecov** [reply]
Great, thanks for clarifying. Closing as the question is answered even though SA doesn't support this use case.
